src/actions/google/analytics/lib/ga_worker.ts

```diff
@@ -0,0 +1,352 @@
+import { Credentials } from "google-auth-library"
+import { analytics_v3, google } from "googleapis"
+import * as Hub from "../../../../hub"
+import { MissingAuthError } from "../../common/missing_auth_error"
+import { MissingRequiredParamsError } from "../../common/missing_required_params_error"
+import { safeParseJson } from "../../common/utils"
+import { GoogleAnalyticsDataImportAction } from "../data_import"
+import { CsvHeaderTransformStream } from "./csv_header_transform_stream"
+
+interface GAUserState {
+  tokens: Credentials
+  redirect: string
+  lastUsedFormParams: any
+}
+
+export class GoogleAnalyticsActionWorker {
+
+  static async fromHubRequest(hubRequest: Hub.ActionRequest, actionInstance: GoogleAnalyticsDataImportAction) {
+    const gaWorker = new GoogleAnalyticsActionWorker(hubRequest, actionInstance)
+    return gaWorker
+  }
+
+  gaClient: analytics_v3.Analytics
+  userState: GAUserState
+  formParams: any
+  newUploadId?: string
+
+  constructor(readonly hubRequest: Hub.ActionRequest, readonly actionInstance: GoogleAnalyticsDataImportAction) {
+    const tmpState = safeParseJson(hubRequest.params.state_json)
+
+    if (!tmpState || !tmpState.tokens || !tmpState.redirect) {
+      throw new MissingAuthError("User state was missing or did not contain tokens & redirect")
+    }
+
+    this.userState = tmpState
+    this.formParams = hubRequest.formParams
+    this.gaClient = this.makeGAClient()
+  }
+
+  log(level: string, ...rest: any[]) {
+    return this.actionInstance.log(level, ...rest)
```

Comment:
This log is not valid in actionhub if it is not using Winston

Reply:
Hey @phillipperalez , I'm not sure I follow. This method redirects to the action instance which in turn uses Winston. Are you saying that it isn't guaranteed Winston will be available, and that actions shouldn't call it directly? I didn't see any other logging interfaces, but could be missing something. I didn't use Winston directly in all the helper code because I wanted to be able to encapsulate some logic like adding the prefix.